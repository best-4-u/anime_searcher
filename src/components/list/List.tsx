import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";

import { RootState } from "../../state/store";
import { IAnimeDetails } from "../../models/animeList/IAnimeListDetails";

import ListItem from "../ListItem/ListItem";

import styles from "./List.module.scss";
import { clearList } from "../../state/slices/anime/animeSlice";

export default function List() {
  const list = useSelector<RootState, IAnimeDetails[]>(
    (state) => state.animeList.list
  );

  const dispatch = useDispatch();

  let navigate = useNavigate();

  useEffect(() => {
    return () => {
      dispatch(clearList());
    };
  }, [dispatch]);

  const onClickItem = useCallback(
    (slug: string) => {
      navigate(`${slug}`);
    },
    [navigate]
  );

  return (
    <div className={styles.anime_list}>
      {list.map((item) => {
        return (
          <ListItem
            key={item.id}
            slug={item.attributes.slug}
            itemDetails={item.attributes}
            clickHandler={onClickItem}
          />
        );
      })}
    </div>
  );
}
