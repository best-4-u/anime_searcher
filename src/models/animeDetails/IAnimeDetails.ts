export interface IFrom {
    day: number;
    month: number;
    year: number;
}

export interface ITo {
    day: number;
    month: number;
    year: number;
}

export interface IProp {
    from: IFrom;
    to: ITo;
}

export interface IAired {
    from: Date;
    to: Date;
    prop: IProp;
    string: string;
}

export interface IAdaptation {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface ISideStory {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface ISummary {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface IRelated {
    Adaptation: IAdaptation[];
    Sidestory: ISideStory[];
    Summary: ISummary[];
}

export interface IProducer {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface ILicensor {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface IStudio {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface IGenre {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface ITheme {
    mal_id: number;
    type: string;
    name: string;
    url: string;
}

export interface IExternalLink {
    name: string;
    url: string;
}

export interface IAnimeDetails {
    request_hash: string;
    request_cached: boolean;
    request_cache_expiry: number;
    mal_id: number;
    url: string;
    image_url: string;
    trailer_url: string;
    title: string;
    title_english: string;
    title_japanese: string;
    title_synonyms: any[];
    type: string;
    source: string;
    episodes: number;
    status: string;
    airing: boolean;
    aired: IAired;
    duration: string;
    rating: string;
    score: number;
    scored_by: number;
    rank: number;
    popularity: number;
    members: number;
    favorites: number;
    synopsis: string;
    background: string;
    premiered: string;
    broadcast: string;
    related: IRelated;
    producers: IProducer[];
    licensors: ILicensor[];
    studios: IStudio[];
    genres: IGenre[];
    explicit_genres: any[];
    demographics: any[];
    themes: ITheme[];
    opening_themes: string[];
    ending_themes: string[];
    external_links: IExternalLink[];
}
