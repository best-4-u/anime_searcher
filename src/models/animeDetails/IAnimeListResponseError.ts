export interface IAnimeListResponseErrorType {
  code: string;
  detail: string;
  status: string;
  title: string;
}

export interface IAnimeListResponseError {
  errors: IAnimeListResponseErrorType[];
}
