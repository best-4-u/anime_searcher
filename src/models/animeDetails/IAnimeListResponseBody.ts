import { IAnimeDetails } from "./IAnimeDetails";

export interface IAnimeListResponse {
  data: IAnimeDetails[];
}
