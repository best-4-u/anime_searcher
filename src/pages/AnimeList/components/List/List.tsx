import { FC, useCallback } from "react";
import { useNavigate } from "react-router-dom";

import { IAnimeDetails } from "../../../../models/animeDetails/IAnimeDetails";

import ListItem from "../ListItem/ListItem";

import styles from "./List.module.scss";

interface props {
  list: IAnimeDetails[]
}

const List: FC<props> = ({ list } : props) => {
  let navigate = useNavigate();

  const onClickItem = useCallback(
    (slug: string) => {
      navigate(`${slug}`);
    },
    [navigate]
  );

  return (
    <div className={styles.anime_list}>
      {list.map((item) => {
        return (
          <ListItem
            key={item.id}
            slug={item.attributes.slug}
            itemDetails={item.attributes}
            clickHandler={onClickItem}
          />
        );
      })}
    </div>
  );
}

export default List;