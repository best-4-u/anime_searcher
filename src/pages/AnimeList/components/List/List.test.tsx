import { create } from "react-test-renderer";
import { IAnimeDetails } from "../../../../models/animeDetails/IAnimeDetails";
// import { list } from "../../../../__tests__/data/data";
import List from "./List";

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("List component", () => {
  const date = new Date();

  const list: IAnimeDetails[] = [
    {
      id: "1",
      type: "anime",
      attributes: {
        createdAt: date,
        updatedAt: date,
        slug: "one-piece",
        synopsis: 'Gol D. Roger was known as the "Pirate King"',
        description: 'Gol D. Roger was known as the "Pirate King"',
        coverImageTopOffset: 50,
        canonicalTitle: "One Piece",
        abbreviatedTitles: ["ワンピース"],
        averageRating: "83.1",
        userCount: 186101,
        favoritesCount: 6578,
        startDate: "1999-10-20",
        endDate: null,
        nextRelease: date,
        popularityRank: 14,
        ratingRank: 33,
        ageRating: "PG",
        ageRatingGuide: "Teens 13 or older",
        subtype: "TV",
        status: "current",
        tba: null,
        posterImage: {
          tiny: "https://media.kitsu.io/anime/poster_images/12/tiny.jpg",
          large: "https://media.kitsu.io/anime/poster_images/12/large.jpg",
          small: "https://media.kitsu.io/anime/poster_images/12/small.jpg",
          medium: "https://media.kitsu.io/anime/poster_images/12/medium.jpg",
          original:
            "https://media.kitsu.io/anime/poster_images/12/original.png",
        },
        coverImage: {
          tiny: "https://media.kitsu.io/anime/12/cover_image/tiny-cd1b0729d5c15400bfa2441ea3751e86.jpeg",
          large:
            "https://media.kitsu.io/anime/12/cover_image/large-3e72f400a87b5241780c5082f0582611.jpeg",
          small:
            "https://media.kitsu.io/anime/12/cover_image/small-8d0cbc39cac65d5d7c4db5f5b3742ae7.jpeg",
          original:
            "https://media.kitsu.io/anime/12/cover_image/21ecb556255bd46b95aea4779d19789f.jpg",
        },
        episodeCount: null,
        episodeLength: 24,
        totalLength: 27960,
        youtubeVideoId: "CmTeYj2FmRc",
        showType: "TV",
        nsfw: false,
      },
    },
    {
      id: "2",
      type: "anime",
      attributes: {
        createdAt: date,
        updatedAt: date,
        slug: "Naruto",
        synopsis: "Naruto is a boy shinobi...",
        description: "Naruto is a boy shinobi...",
        coverImageTopOffset: 50,
        canonicalTitle: "Naruto",
        abbreviatedTitles: ["ワンピース"],
        averageRating: "97.1",
        userCount: 186101,
        favoritesCount: 6578,
        startDate: "1995-10-20",
        endDate: null,
        nextRelease: date,
        popularityRank: 2,
        ratingRank: 1,
        ageRating: "PG",
        ageRatingGuide: "Teens 12 or older",
        subtype: "TV",
        status: "current",
        tba: null,
        posterImage: {
          tiny: "https://media.kitsu.io/anime/poster_images/12/tiny.jpg",
          large: "https://media.kitsu.io/anime/poster_images/12/large.jpg",
          small: "https://media.kitsu.io/anime/poster_images/12/small.jpg",
          medium: "https://media.kitsu.io/anime/poster_images/12/medium.jpg",
          original:
            "https://media.kitsu.io/anime/poster_images/12/original.png",
        },
        coverImage: {
          tiny: "https://media.kitsu.io/anime/12/cover_image/tiny-cd1b0729d5c15400bfa2441ea3751e86.jpeg",
          large:
            "https://media.kitsu.io/anime/12/cover_image/large-3e72f400a87b5241780c5082f0582611.jpeg",
          small:
            "https://media.kitsu.io/anime/12/cover_image/small-8d0cbc39cac65d5d7c4db5f5b3742ae7.jpeg",
          original:
            "https://media.kitsu.io/anime/12/cover_image/21ecb556255bd46b95aea4779d19789f.jpg",
        },
        episodeCount: null,
        episodeLength: 24,
        totalLength: 27960,
        youtubeVideoId: "CmTeYj2FmRc",
        showType: "TV",
        nsfw: false,
      },
    },
  ];

  it("shows component with 2 items correctly", () => {
    const tree = create(<List list={list} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
