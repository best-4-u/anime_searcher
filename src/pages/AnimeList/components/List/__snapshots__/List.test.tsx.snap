// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component shows component with 2 items correctly 1`] = `
<div
  className="anime_list"
>
  <div
    className="list_item"
  >
    <div
      className="left_side"
    >
      <img
        alt="anime"
        src="https://media.kitsu.io/anime/poster_images/12/medium.jpg"
      />
    </div>
    <div
      className="right_side"
    >
      <div
        className="anime_item_title"
      >
        <span
          onClick={[Function]}
          role="button"
        >
          One Piece
        </span>
      </div>
      <div
        className="anime_item_extra"
      >
        <div>
          <span>
             Rating: 
          </span>
          <span>
             
            83.1
             
          </span>
        </div>
        <div>
          <span>
             Type: 
          </span>
          <span>
             
            TV
             
          </span>
        </div>
        <div>
          <span>
             Episodes: 
          </span>
          <span>
             
            Unknown
             
          </span>
        </div>
      </div>
      <div>
        <div
          className="clamp"
        >
          Gol D. Roger was known as the "Pirate King"
        </div>
        <span
          className="clamp_button"
          onClick={[Function]}
        >
          Read 
          more
        </span>
      </div>
    </div>
  </div>
  <div
    className="list_item"
  >
    <div
      className="left_side"
    >
      <img
        alt="anime"
        src="https://media.kitsu.io/anime/poster_images/12/medium.jpg"
      />
    </div>
    <div
      className="right_side"
    >
      <div
        className="anime_item_title"
      >
        <span
          onClick={[Function]}
          role="button"
        >
          Naruto
        </span>
      </div>
      <div
        className="anime_item_extra"
      >
        <div>
          <span>
             Rating: 
          </span>
          <span>
             
            97.1
             
          </span>
        </div>
        <div>
          <span>
             Type: 
          </span>
          <span>
             
            TV
             
          </span>
        </div>
        <div>
          <span>
             Episodes: 
          </span>
          <span>
             
            Unknown
             
          </span>
        </div>
      </div>
      <div>
        <div
          className="clamp"
        >
          Naruto is a boy shinobi...
        </div>
        <span
          className="clamp_button"
          onClick={[Function]}
        >
          Read 
          more
        </span>
      </div>
    </div>
  </div>
</div>
`;
